File: ios/Chirp/Chirp/Components/NostrContent/ContentTreeWire.swift
import SwiftUI

/// Swift Codable mirror of the Rust `nmp_content::wire::ContentTreeWire`.
///
/// The Rust type is a flat *arena*: `nodes: [WireNode]` plus `roots: [UInt32]`.
/// Every recursive parent→child edge in the source tree is a `[UInt32]` of
/// indices into `nodes`. See `crates/nmp-content/src/wire/mod.rs` for the
/// canonical definition.
///
/// This file is shipped as a registry component so apps that install
/// `swiftui/content-core` get a complete, drift-resistant mirror without
/// hand-rolling Decodables.
public struct ContentTreeWire: Decodable, Equatable, Sendable {
    public let nodes: [NostrWireNode]
    public let roots: [UInt32]
    public let mode: String?

    public init(nodes: [NostrWireNode], roots: [UInt32], mode: String? = nil) {
        self.nodes = nodes
        self.roots = roots
        self.mode = mode
    }

    /// Bounds-checked arena lookup.
    public func node(at index: UInt32) -> NostrWireNode? {
        let i = Int(index)
        guard i >= 0, i < nodes.count else { return nil }
        return nodes[i]
    }
}

/// Typed mirror of the Rust `nmp_content::MediaKind` enum. Raw values match the
/// PascalCase Rust serde representation (the Rust enum has no `rename_all`).
public enum NostrMediaKind: String, Decodable, Equatable, Sendable {
    case image = "Image"
    case video = "Video"
    case audio = "Audio"
}

/// NIP-21 entity discriminator on the wire.
public enum NostrWireUriKind: String, Decodable, Equatable, Sendable {
    case profile
    case event
    case address
}

/// Why a `placeholder` wire node was emitted.
public enum NostrWirePlaceholderReason: String, Decodable, Equatable, Sendable {
    case depthLimit = "depth_limit"
    case unresolvedUri = "unresolved_uri"
}

/// Reserved payment segment (`WireNode::Invoice`).
public enum NostrWireInvoice: Decodable, Equatable, Sendable {
    case bolt11(String)
    case bolt12(String)
    case cashu(String)

    private enum CodingKeys: String, CodingKey {
        case bolt11 = "Bolt11"
        case bolt12 = "Bolt12"
        case cashu = "Cashu"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let value = try container.decodeIfPresent(String.self, forKey: .bolt11) {
            self = .bolt11(value)
        } else if let value = try container.decodeIfPresent(String.self, forKey: .bolt12) {
            self = .bolt12(value)
        } else if let value = try container.decodeIfPresent(String.self, forKey: .cashu) {
            self = .cashu(value)
        } else {
            throw DecodingError.dataCorrupted(
                .init(codingPath: decoder.codingPath, debugDescription: "unknown invoice kind")
            )
        }
    }
}

/// Flattened, Codable projection of `nmp_core::nip21::NostrUri`. `uri` is the
/// round-trippable canonical `nostr:` URI; `primaryId` is the hex pubkey for
/// profiles, event id for events, or author pubkey for addresses.
public struct NostrWireUri: Decodable, Equatable, Sendable {
    public let uri: String
    public let kind: NostrWireUriKind
    public let primaryId: String
    public let relays: [String]
    public let author: String?
    public let eventKind: UInt32?

    private enum CodingKeys: String, CodingKey {
        case uri, kind, relays, author
        case primaryId = "primary_id"
        case eventKind = "event_kind"
    }

    public init(
        uri: String,
        kind: NostrWireUriKind,
        primaryId: String,
        relays: [String] = [],
        author: String? = nil,
        eventKind: UInt32? = nil
    ) {
        self.uri = uri
        self.kind = kind
        self.primaryId = primaryId
        self.relays = relays
        self.author = author
        self.eventKind = eventKind
    }
}

/// One node in the `ContentTreeWire` arena. Covers every variant of the Rust
/// `WireNode` enum.
public enum NostrWireNode: Decodable, Equatable, Sendable {
    case text(String)
    case mention(NostrWireUri)
    case eventRef(NostrWireUri)
    case hashtag(String)
    case url(String)
    case media(urls: [String], kind: NostrMediaKind)
    case emoji(shortcode: String, url: String?)
    case invoice(NostrWireInvoice)
    case heading(level: UInt8, children: [UInt32])
    case paragraph(children: [UInt32])
    case blockQuote(children: [UInt32])
    case codeBlock(info: String?, body: String)
    case list(orderedStart: UInt64?, items: [[UInt32]])
    case rule
    case emphasis(children: [UInt32])
    case strong(children: [UInt32])
    case inlineCode(String)
    case link(children: [UInt32], href: String?)
    case image(alt: String, title: String?, src: String?)
    case softBreak
    case hardBreak
    case placeholder(reason: NostrWirePlaceholderReason)

    private enum CodingKeys: String, CodingKey {
        case kind, text, uri, tag, url, urls
        case mediaKind = "media_kind"
        case shortcode, level, children, code, info, body, href
        case orderedStart = "ordered_start"
        case items, alt, title, src, reason, invoice
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let kind = try container.decode(String.self, forKey: .kind)
        switch kind {
        case "text":
            self = .text(try container.decode(String.self, forKey: .text))
        case "mention":
            self = .mention(try container.decode(NostrWireUri.self, forKey: .uri))
        case "event_ref":
            self = .eventRef(try container.decode(NostrWireUri.self, forKey: .uri))
        case "hashtag":
            self = .hashtag(try container.decode(String.self, forKey: .tag))
        case "url":
            self = .url(try container.decode(String.self, forKey: .url))
        case "media":
            self = .media(
                urls: try container.decode([String].self, forKey: .urls),
                kind: try container.decode(NostrMediaKind.self, forKey: .mediaKind)
            )
        case "emoji":
            self = .emoji(
                shortcode: try container.decode(String.self, forKey: .shortcode),
                url: try container.decodeIfPresent(String.self, forKey: .url)
            )
        case "invoice":
            self = .invoice(try container.decode(NostrWireInvoice.self, forKey: .invoice))
        case "heading":
            self = .heading(
                level: try container.decode(UInt8.self, forKey: .level),
                children: try container.decode([UInt32].self, forKey: .children)
            )
        case "paragraph":
            self = .paragraph(children: try container.decode([UInt32].self, forKey: .children))
        case "block_quote":
            self = .blockQuote(children: try container.decode([UInt32].self, forKey: .children))
        case "code_block":
            self = .codeBlock(
                info: try container.decodeIfPresent(String.self, forKey: .info),
                body: try container.decode(String.self, forKey: .body)
            )
        case "list":
            self = .list(
                orderedStart: try container.decodeIfPresent(UInt64.self, forKey: .orderedStart),
                items: try container.decode([[UInt32]].self, forKey: .items)
            )
        case "rule":
            self = .rule
        case "emphasis":
            self = .emphasis(children: try container.decode([UInt32].self, forKey: .children))
        case "strong":
            self = .strong(children: try container.decode([UInt32].self, forKey: .children))
        case "inline_code":
            self = .inlineCode(try container.decode(String.self, forKey: .code))
        case "link":
            self = .link(
                children: try container.decode([UInt32].self, forKey: .children),
                href: try container.decodeIfPresent(String.self, forKey: .href)
            )
        case "image":
            self = .image(
                alt: try container.decode(String.self, forKey: .alt),
                title: try container.decodeIfPresent(String.self, forKey: .title),
                src: try container.decodeIfPresent(String.self, forKey: .src)
            )
        case "soft_break":
            self = .softBreak
        case "hard_break":
            self = .hardBreak
        case "placeholder":
            self = .placeholder(
                reason: try container.decode(NostrWirePlaceholderReason.self, forKey: .reason)
            )
        default:
            // Forward-compat: any unknown kind collapses to a depth_limit
            // placeholder rather than failing the whole decode.
            self = .placeholder(reason: .depthLimit)
        }
    }
}

/// Deterministic identicon color for a hex pubkey. Apps that don't supply an
/// avatar URL can render this as a circle background. Ported from Chirp's
/// djb2-based palette so installed apps stay visually consistent.
public enum NostrIdenticon {
    /// Returns a stable `Color` derived from a hex pubkey (or any string).
    /// Uses the djb2 hash mapped to HSL with fixed S/L for legibility.
    public static func color(forPubkey pubkey: String) -> Color {
        let hue = Double(djb2(pubkey) % 360) / 360.0
        return Color(hue: hue, saturation: 0.55, brightness: 0.75)
    }

    /// Two-character monogram derived from a hex pubkey (first two chars,
    /// uppercased). Apps with kind:0 profile data should provide their own.
    public static func initials(forPubkey pubkey: String) -> String {
        let trimmed = pubkey.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return "?" }
        return String(trimmed.prefix(2)).uppercased()
    }

    private static func djb2(_ value: String) -> UInt32 {
        var hash: UInt32 = 5381
        for byte in value.utf8 {
            hash = (hash &* 33) &+ UInt32(byte)
        }
        return hash
    }
}

File: apps/nmp-gallery/android/app/src/main/kotlin/org/nmp/gallery/registry/ContentTreeWire.kt
// Requires: kotlinx-serialization-json (>= 1.6), Kotlin 1.9+
//
// Kotlin mirror of the Rust `nmp_content::wire::ContentTreeWire`. See
// `crates/nmp-content/src/wire/mod.rs` for the canonical definition.
//
// The Rust type is a flat *arena*: `nodes: Vec<WireNode>` plus
// `roots: Vec<u32>`. Every recursive parent → child edge in the source tree is
// a `Vec<u32>` of indices into `nodes`. This file is shipped as a registry
// component so apps that install `compose/content-core` get a complete,
// drift-resistant mirror without hand-rolling Decodable equivalents.
//
// Discriminator strategy: `WireNode` uses `@JsonClassDiscriminator("kind")` to
// match the Rust `#[serde(tag = "kind", rename_all = "snake_case")]`. Apps
// configuring their own `Json` instance must keep `classDiscriminator = "kind"`
// (or pass `useArrayPolymorphism = false`, the default).

package org.nmp.gallery.registry

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp

@Serializable
public data class ContentTreeWire(
    /** Flat arena of every node in the tree (block + inline kinds). */
    val nodes: List<WireNode> = emptyList(),
    /** Top-level node indices, in document order. */
    val roots: List<UInt> = emptyList(),
    /** The mode the source tree was produced under (mirrors `RenderMode`). */
    val mode: String? = null,
) {
    /** Bounds-checked arena lookup. Returns `null` for out-of-range indices. */
    public fun nodeAt(index: UInt): WireNode? {
        val i = index.toLong()
        if (i < 0 || i >= nodes.size.toLong()) return null
        return nodes[i.toInt()]
    }
}

/**
 * Typed mirror of the Rust `nmp_content::MediaKind` enum. Serialized values
 * match the PascalCase Rust serde representation (the Rust enum has no
 * `rename_all`).
 */
@Serializable
public enum class MediaKind {
    @SerialName("Image") Image,
    @SerialName("Video") Video,
    @SerialName("Audio") Audio,
}

/** NIP-21 entity discriminator on the wire. */
@Serializable
public enum class WireNostrUriKind {
    @SerialName("profile") Profile,
    @SerialName("event") Event,
    @SerialName("address") Address,
}

/** Why a `placeholder` wire node was emitted. */
@Serializable
public enum class PlaceholderReason {
    @SerialName("depth_limit") DepthLimit,
    @SerialName("unresolved_uri") UnresolvedUri,
}

/**
 * Reserved payment segment (`WireNode::Invoice`).
 *
 * The Rust `InvoiceKind` enum has bare `Serialize`/`Deserialize` derives
 * (no `#[serde(tag = "…")]`), so serde emits the default
 * **externally-tagged** form — one of:
 *
 *     {"Bolt11": "lnbc…"}
 *     {"Bolt12": "lno…"}
 *     {"Cashu":  "cashuA…"}
 *
 * We mirror that on the wire as three mutually-exclusive optional fields
 * rather than a `sealed class`, because `kotlinx.serialization`'s default
 * polymorphic encoding for sealed classes is *internally* tagged (`"type"`
 * key + payload fields at the same level), which would not decode the
 * externally-tagged JSON Rust actually emits. Exactly one of the three
 * fields will be present in a well-formed payload.
 */
@Serializable
public data class WireInvoice(
    @SerialName("Bolt11") val bolt11: String? = null,
    @SerialName("Bolt12") val bolt12: String? = null,
    @SerialName("Cashu") val cashu: String? = null,
)

/**
 * Flattened, Serializable projection of `nmp_core::nip21::NostrUri`. `uri` is
 * the round-trippable canonical `nostr:` URI; `primaryId` is the hex pubkey
 * for profiles, event id for events, or author pubkey for addresses.
 */
@Serializable
public data class WireNostrUri(
    val uri: String,
    val kind: WireNostrUriKind,
    @SerialName("primary_id") val primaryId: String,
    val relays: List<String> = emptyList(),
    val author: String? = null,
    @SerialName("event_kind") val eventKind: UInt? = null,
)

/**
 * One node in the [ContentTreeWire] arena. Covers every variant of the Rust
 * `WireNode` enum. The class discriminator field is `kind` to match the Rust
 * `#[serde(tag = "kind")]` representation.
 */
@OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("kind")
public sealed class WireNode {
    /** Inline text run. */
    @Serializable
    @SerialName("text")
    public data class Text(val text: String) : WireNode()

    /** Profile mention. */
    @Serializable
    @SerialName("mention")
    public data class Mention(val uri: WireNostrUri) : WireNode()

    /** Event / address reference. */
    @Serializable
    @SerialName("event_ref")
    public data class EventRef(val uri: WireNostrUri) : WireNode()

    /** `#hashtag` without leading `#`. */
    @Serializable
    @SerialName("hashtag")
    public data class Hashtag(val tag: String) : WireNode()

    /** Plain URL. */
    @Serializable
    @SerialName("url")
    public data class Url(val url: String) : WireNode()

    /** Ad-candidate URL. Renders exactly like [Url] — a plain link. */
    @Serializable
    @SerialName("ad_candidate_url")
    public data class AdCandidateUrl(val url: String) : WireNode()

    /** Grouped media block. */
    @Serializable
    @SerialName("media")
    public data class Media(
        val urls: List<String>,
        @SerialName("media_kind") val mediaKind: MediaKind,
    ) : WireNode()

    /** NIP-30 custom emoji. */
    @Serializable
    @SerialName("emoji")
    public data class Emoji(
        val shortcode: String,
        val url: String? = null,
    ) : WireNode()

    /** Reserved payment segment. */
    @Serializable
    @SerialName("invoice")
    public data class Invoice(val invoice: WireInvoice) : WireNode()

    /** Markdown heading. */
    @Serializable
    @SerialName("heading")
    public data class Heading(
        val level: UByte,
        val children: List<UInt>,
    ) : WireNode()

    /** Markdown paragraph. */
    @Serializable
    @SerialName("paragraph")
    public data class Paragraph(val children: List<UInt>) : WireNode()

    /** Markdown block quote. */
    @Serializable
    @SerialName("block_quote")
    public data class BlockQuote(val children: List<UInt>) : WireNode()

    /** Markdown fenced/indented code block (verbatim, never tokenized). */
    @Serializable
    @SerialName("code_block")
    public data class CodeBlock(
        val info: String? = null,
        val body: String,
    ) : WireNode()

    /** Markdown bullet/ordered list. */
    @Serializable
    @SerialName("list")
    public data class ListNode(
        @SerialName("ordered_start") val orderedStart: ULong? = null,
        val items: List<List<UInt>>,
    ) : WireNode()

    /** Markdown horizontal rule. */
    @Serializable
    @SerialName("rule")
    public object Rule : WireNode()

    /** `*italic*` — children are inline node indices. */
    @Serializable
    @SerialName("emphasis")
    public data class Emphasis(val children: List<UInt>) : WireNode()

    /** `**bold**` — children are inline node indices. */
    @Serializable
    @SerialName("strong")
    public data class Strong(val children: List<UInt>) : WireNode()

    /** Inline `` `code` `` (verbatim). */
    @Serializable
    @SerialName("inline_code")
    public data class InlineCode(val code: String) : WireNode()

    /** `[label](href)` — `label` children are inline node indices. */
    @Serializable
    @SerialName("link")
    public data class Link(
        val children: List<UInt>,
        val href: String? = null,
    ) : WireNode()

    /** `![alt](src "title")`. */
    @Serializable
    @SerialName("image")
    public data class Image(
        val alt: String,
        val title: String? = null,
        val src: String? = null,
    ) : WireNode()

    /** Soft line break. */
    @Serializable
    @SerialName("soft_break")
    public object SoftBreak : WireNode()

    /** Hard line break. */
    @Serializable
    @SerialName("hard_break")
    public object HardBreak : WireNode()

    /**
     * Typed placeholder: content existed here but could not be projected.
     * Never a dropped subtree — always a renderable node.
     */
    @Serializable
    @SerialName("placeholder")
    public data class Placeholder(val reason: PlaceholderReason) : WireNode()
}

/**
 * Deterministic identicon helpers for a hex pubkey.
 *
 * Algorithm is the 5×5 symmetric djb2 grid (GitHub-style): the lower 15 bits
 * of the djb2 hash determine which of the 15 left-half cells (3 cols × 5 rows)
 * are filled; columns 3–4 mirror columns 1–0 so the result is horizontally
 * symmetric. Color uses djb2 % 360 as the HSV hue with S=0.55, V=0.75.
 *
 * This algorithm is byte-identical to the Swift implementation in
 * `swiftui/content-core/ContentTreeWire.swift` and
 * `swiftui/user-avatar/NostrAvatar.swift`. Same pubkey → same grid and color
 * on every platform.
 */
public object NostrIdenticon {
    /** Stable HSV [Color] derived from a hex pubkey (or any string).
     *  Uses djb2 hash mapped to hue with fixed S=0.55, V=0.75 for legibility.
     */
    public fun colorForPubkey(pubkey: String): Color {
        val hue = (djb2(pubkey) % 360u).toFloat()
        return Color.hsv(hue, 0.55f, 0.75f)
    }

    /**
     * Returns a 5×5 symmetric fill pattern for a pubkey. Row-major: `cells[row][col]`
     * is `true` iff that cell is filled. The left 3 columns (0–2) are determined by
     * the lower 15 bits of djb2; columns 3–4 mirror columns 1–0.
     */
    public fun cellsForPubkey(pubkey: String): Array<BooleanArray> {
        val hash = djb2(pubkey)
        return Array(5) { row ->
            BooleanArray(5) { col ->
                val mirrorCol = if (col < 3) col else 4 - col
                val bit = row * 3 + mirrorCol
                (hash shr bit) and 1u == 1u
            }
        }
    }

    /**
     * Two-character monogram derived from a hex pubkey (first two chars,
     * uppercased). Apps with kind:0 profile data should provide their own.
     */
    public fun initialsForPubkey(pubkey: String): String {
        val trimmed = pubkey.trim()
        if (trimmed.isEmpty()) return "?"
        return trimmed.take(2).uppercase()
    }

    private fun djb2(value: String): UInt {
        var hash: UInt = 5381u
        for (byte in value.encodeToByteArray()) {
            hash = hash * 33u + byte.toUByte().toUInt()
        }
        return hash
    }
}

/**
 * Renders the 5×5 symmetric identicon grid for [pubkey] using [NostrIdenticon].
 * Filled cells are drawn in the pubkey color; empty cells show the same color
 * at 15% opacity so the tile reads as a tinted patch rather than blank.
 */
@Composable
public fun NostrIdenticonGrid(pubkey: String, size: Dp, modifier: Modifier = Modifier) {
    val color = NostrIdenticon.colorForPubkey(pubkey)
    val cells = remember(pubkey) { NostrIdenticon.cellsForPubkey(pubkey) }
    Canvas(
        modifier = modifier
            .size(size)
            .background(color.copy(alpha = 0.15f)),
    ) {
        val spacing = 1f
        val cellPx = (minOf(this.size.width, this.size.height) - spacing * 4f) / 5f
        for (row in 0..4) {
            for (col in 0..4) {
                if (cells[row][col]) {
                    drawRect(
                        color = color,
                        topLeft = Offset(col * (cellPx + spacing), row * (cellPx + spacing)),
                        size = Size(cellPx, cellPx),
                    )
                }
            }
        }
    }
}
